_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+saturate available CPU, but Rust does not, yet.
+
+The latency "cutoff" at the very right of the graph corresponds with the change of mode:
+from saturation-induced latency to appearance of error responses.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage.svg" />
+
+Here we measure high-water mark memory usage of the container from its start till the end of each
+benchmark step, as reported by Docker (i.e. not just momentary memory usage).
+
+The comparison between JVM-based and Rust frameworks is not representative here
+as allocated Java heap depends on configuration in addition to actual usage.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage_per_requests.svg" />
+
+The same metric as above, but divided by the number of successful requests per second;
+which gives an unconventional unit of *megabyte-seconds per request*.
+
+It is normal that the values for low connections counts are higher
+because the base memory footprint of each framework is spread out across fewer requests.
+
+You may notice a slight difference in shape between Http4k and the 2 async frameworks:
+Http4k gets down to ~0.05 MiB⋅s and stays there, while the others go up again as latencies increase.
+If you have an explanation for this, please share it in a discussion.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu.svg" />
+
+Consumed CPU time as reported by Docker API.
+
+JVM is seen working hard to JIT-optimise Kotlin-based codebases during the warm-up steps.
+The just-in-time optimisation, combined with the execution of unoptimised code, takes roughly 23 CPU-seconds extra,
+a remarkably small amount of resources.[^jit]
+
+[^jit]: It is tempting to make an unfair comparison to Rust,
+    where optimised build adds 17 CPU-*minutes* on top of 11.5 CPU-minute debug build.
+    We should also acknowledge the artificially low amount of actually executed Kotlin application code in our benchmark.
+
+Still, this is something to keep in mind about JVM-based microservices
+when considering, for example, high-frequency continuous deployment or aggressive instance count autoscaling.
+The JIT may compete for resources with the not-yet-optimised target workload.
+Such characteristic should also apply to other JIT-based engines like Node.js' V8.
+
+Further right in the graph, Kotlin implementations quickly saturate the allocated CPU portion,
+Rust following ~2 steps later.
+
+I'm curious that Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu_per_request.svg" />
+
+Consumed CPU time divided by successful requests per second.
+
+I'm not sure how to explain the improving efficiency of Rust (Actix) with the increasing number of concurrent connections.
```

Comment:
Followup blogpost? 😉

Reply:
Maybe, but I would have to do quite a lot of research for that. Let's see if the posts atracts some comments from more knowledgeable.

Reply:
🤞